[JENKINS-26574] Allow distinct template for matrix head build

lavaliere opened the pull request:
[JENKINS-26574](https://issues.jenkins-ci.org/browse/JENKINS-26574) - Added support for Matrix projects by adding an extra field and hiding this field if the project type is not Matrix. There are 3 cases I can think of when it comes to these fields:

* template set, matrix empty (presumably != matrix build, ignore matrix field values)
* matrix set, template empty (presumably matrix build with child builds taking the same name pattern as the parent build)
* template and matrix set - use case that a CloudBees customer was asking for. Allows Matrix build name to be set explicitly and separately from spawned builds in the case where spawned builds take their name from a build-generated file.

I also added a test cases to each method to check that both case 2 and case 3 above are being set correctly.

lavaliere commented:
@reviewbybees

daniel-beck commented:
Commits should probably be squashed.

daniel-beck commented on src/main/java/org/jenkinsci/plugins/buildnamesetter/BuildNameSetter.java:
```diff
@@ -26,48 +28,106 @@
             public boolean endBuild() throws InterruptedException, IOException {
-                setDisplayName(build,listener);
-                return super.endBuild();
+            	try{
```
Inconsistent coding style (see line 108).

daniel-beck commented on src/main/java/org/jenkinsci/plugins/buildnamesetter/BuildNameSetter.java:
```diff
@@ -26,48 +28,106 @@
+                try {
+					setDisplayName(build,listener);
+					started = true;
+	                return super.startBuild();
```
Indentation is a mess here.

recampbell replied:
This is because tabs are used instead of spaces.

daniel-beck commented on src/main/java/org/jenkinsci/plugins/buildnamesetter/BuildNameSetter.java:
```diff
@@ -5,11 +5,13 @@
 import hudson.matrix.MatrixAggregatable;
 import hudson.matrix.MatrixAggregator;
 import hudson.matrix.MatrixBuild;
+import hudson.matrix.MatrixProject;
```
While the plugin is currently on 1.480.x, what happens to this dependency when that baseline gets raised? Since Matrix Plugin has been detached from core a while back, is there a way for this PR to be implemented that allows the Matrix Plugin dependency to be made optional?

daniel-beck replied:
Correction: This change will probably break the plugin for anyone having disabled Matrix Project.

daniel-beck commented on src/test/java/org/jenkinsci/plugins/buildnamesetter/BuildNameSetterTest.java:
```diff
@@ -1,72 +1,156 @@
-package org.jenkinsci.plugins.buildnamesetter;
```
I don't know what you did here (line breaks change from Windows/Unix to Unix/Windows), but the diff shouldn't look like this.

daniel-beck commented on src/main/resources/org/jenkinsci/plugins/buildnamesetter/BuildNameSetter/help-template.jelly:
```diff
@@ -3,4 +3,13 @@
+    For multi-configuration projects: if the Matrix Build Name is set and the Build 
+    Name field is not set, then the head build and child builds will all use the Matrix Build 
+    Name's pattern.
+    
```
Not actually a paragraph break in the rendered HTML output.

recampbell commented on src/main/java/org/jenkinsci/plugins/buildnamesetter/BuildNameSetter.java:
```diff
@@ -26,48 +28,106 @@
+        matches = template.equals(matrixTemplate);
+        
+        if(matches || matrixTemplate.length()>0 && template.length()==0){
+        	scenario = 0;
```
Please use 4 spaces instead of tabs when indenting.

daniel-beck commented on src/main/java/org/jenkinsci/plugins/buildnamesetter/BuildNameSetter.java:
```diff
@@ -26,48 +28,106 @@
             }
         };
     }
+   
```
Try to avoid unnecessary whitespace changes in existing code.

daniel-beck commented on src/main/java/org/jenkinsci/plugins/buildnamesetter/BuildNameSetter.java:
```diff
@@ -26,48 +28,106 @@
+        	scenario = 0;
+        }else if (matrixTemplate.length()>0 && template.length()>0){
+        	scenario = 1;
+        }else if (matrixTemplate.length()==0 && template.length()>0){
```
The plugin's current coding style is to add spaces before open curly braces and after closing curly braces, so this should probably be made consistent.

christ66 commented on src/main/java/org/jenkinsci/plugins/buildnamesetter/BuildNameSetter.java:
```diff
@@ -26,48 +28,106 @@
+        	//not checking before setting name because this check is done via Jelly call to isMatrix()
+        	setDisplayName(build, listener);
+        }catch(Exception e){
+        	e.printStackTrace();
```
Should be `e.printStackTrace(listener.getLogger())` otherwise the stack trace and error message are lost to standard out/err. Also the catch block should be changed to only catch `IllegalArgumentException`.

christ66 commented on src/main/java/org/jenkinsci/plugins/buildnamesetter/BuildNameSetter.java:
```diff
@@ -26,48 +28,106 @@
+	    			build.setDisplayName(TokenMacro.expandAll(build, listener, template));
+    				return;
+    		case -1:
+    				throw new Exception("No valid values set for build names!");
```
Should probably not be throwing a new `Exception`, but instead `java.lang.IllegalArgumentException`. Then the `private void setDisplayName(AbstractBuild build, BuildListener listener) throws Exception` should be changed to throw `IllegalArgumentException`.

stephenc commented:
@daniel-beck I disagree, as long as the merge is not fast-forwarded. If you are going to fast-forward the merge then yes the commits *must* be squashed, but otherwise it can be good to see the thought processes of the developer

jglick commented:
> If you are going to fast-forward the merge

Note that the _Merge pull request_ button passes the equivalent of `--no-ff` so this is usually not a concern. I would also discourage use of history rewriting (except in special cases such as large binaries being inadvertently committed).

lavaliere commented:
@reviewbybees - attempted to make all of the selected changes re: styling (switched from Eclipse to NetBeans), exceptions, and making the matrix-project optional

stephenc commented on pom.xml:
```diff
@@ -3,7 +3,7 @@
     <groupId>org.jenkins-ci.plugins</groupId>
     <artifactId>plugin</artifactId>
-    <version>1.480.3</version>
+    <version>1.599</version>
```
Really? can you not live with even the most recent LTS?

I would think you could even go as far back as 1.565.1

lavaliere commented:
@stephenc Point taken. 

I changed the required version to 1.561, which is the minimum required for the version of the Matrix plugin whose methods we call in this code. I had changed it to 1.599 on the suggestion of Jesse's NetBeans Jenkins plugin. 

Perhaps leaving 1.480 as the required version would be preferred though, since the Matrix builds calls are conditional on the plugin being installed at all...?

alecharp commented:
@lavaliere - I think 1.561 is good because of the Matrix plugin version used (1.4). Letting the 1.480.x version could lead to other issues which would be complicated to fix on 1.480.x base.

daniel-beck commented:
It appears the current plugin already has a dependency on matrix projects (as it currently just sets the parent build's description to the same string), I just misread the earlier change. Sorry!

Making the dependency optional is probably misleading; what happens when you run this with Matrix Project Plugin disabled? I'd expect it to fail to load the publisher as it is (and was) `MatrixAggregatable`.

(This is what happens when dependencies to detached plugins are implied, so raising the base version to 1.561-1.565.x would make the dependency explicit at least, without excluding a majority of users.)

Using LTS versions as plugin base versions is preferred, see https://wiki.jenkins-ci.org/display/JENKINS/Plugin+Report+Card.

ndeloof commented:
see https://github.com/jenkinsci/artifactdeployer-plugin/pull/8 as an attempt to make matrix-plugin an optional dependency but address the `MatrixAggregatable` issue

jglick commented:
> I had changed it to 1.599 on the suggestion of Jesse's NetBeans Jenkins plugin.

For the record: my plugin does not suggest changes to the baseline version. The generic Maven support does suggest using the latest release version of a parent POM if a newer one is known from the Maven index. In this case the suggestion should generally be ignored.

jglick commented on src/main/java/org/jenkinsci/plugins/buildnamesetter/BuildNameSetter.java:
```diff
@@ -26,50 +28,106 @@
+   
 
-    @Extension
+    @Extension(optional = true)
```
As noted elsewhere, this is just wrong; you cannot make the dependency on `matrix-project` optional this way. Currently I think there is _no_ reasonable way to make it optional.

lavaliere replied:
Okay. I got that tag from the OSS wiki [on plugin dependencies](https://wiki.jenkins-ci.org/display/JENKINS/Dependencies+among+plugins): 

> You can mark your extension as an optional extension by changing the @Extension annotation to @Extension(optional = true) and Jenkins don't log any class loading errors when reading it.

I'll look through the options [on this page](https://wiki.jenkins-ci.org/display/JENKINS/Tips+for+optional+dependencies) instead.

jglick replied:
Yes `Extension.optional` exists and yes it sometimes works, but you have to know exactly what you are doing. Basically all it does is tell Jenkins that if _when loading the extension_ a `LinkageError` (typically `NoClassDefFoundError`) is thrown, to quietly skip the extension rather than print a big stack trace. For some simple cases, like extensions that just implement an SPI in some other plugin, that is enough. But making the dependency optional here is much more complicated than that.

jglick replied:
As noted above, I do not think it is feasible for the dependency to be in any way optional currently. Would really require a new API in `matrix-project`. Anyway optional dependencies are always a headache; prefer to just keep it as a hard dep.

stephenc commented:
Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/jenkinsci/build-name-setter-plugin/4)

Reviewed files:
- pom.xml @ r3
- src/main/java/org/jenkinsci/plugins/buildnamesetter/BuildNameSetter.java @ r2
- src/main/resources/org/jenkinsci/plugins/buildnamesetter/BuildNameSetter/config.jelly @ r1
- src/main/resources/org/jenkinsci/plugins/buildnamesetter/BuildNameSetter/help-template.jelly @ r1
- src/test/java/org/jenkinsci/plugins/buildnamesetter/BuildNameSetterTest.java @ r1

---

Lots of excessive whitespace changes

---

<sup>**[src/main/java/org/jenkinsci/plugins/buildnamesetter/BuildNameSetter.java, line 37 \[r3\]](https://reviewable.io:443/reviews/jenkinsci/build-name-setter-plugin/4#-JlMu_8O6voigHK1JsvJ)** ([raw file](https://github.com/jenkinsci/build-name-setter-plugin/blob/fe06a7abae8571f8426c645fb8e091c9289ee35b/src/main/java/org/jenkinsci/plugins/buildnamesetter/BuildNameSetter.java#L37)):</sup>
gratuitous whitespace changes

---


<!-- Sent from Reviewable.io -->

reviewbybees commented:
This pull request originates from a [CloudBees](https://www.cloudbees.com/) employee. At CloudBees, we require that all pull requests be reviewed by other CloudBees employees before we seek to have the change accepted. If you want to find out more about our process please see [this note](https://github.com/reviewbybees/about)

stephenc commented:
:bug: as there are a lot of outstanding comments remaining here.

ebrahim-moshaya commented:
This update would be really useful!! Any chance for a merge?

lavaliere commented:
Sorry, it's been so long since I wrote this that I had assumed it had been merged. I'll review the comments and update my PR this weekend.

ebrahim-moshaya commented:
@lavaliere thanks for the update, look forward to this getting merged

apemberton commented:
Has this been merged?

oleg-nenashev commented:
@apemberton 
The PR is not ready for merge, CC @lavaliere

Le0Michine commented:
Is the PR alive?

daniel-beck commented:
@lavaliere Ping!

Le0Michine commented:
@lavaliere please respond, I'm going to close pull request if it is not needed more and doesn't ready to merge

oleg-nenashev commented:
I've pinged @lavaliere directly. If there is no response within 24 hours, feel free to close it (you are eligible to do it as a maintainer). BTW I recommend to just add the "needs-fix" label and forget about it. If somebody wants to finish it, he will be able to use the current stuff as a source
